xarray/indexes/interval_index.py

```diff
@@ -0,0 +1,185 @@
+
+        return cls(index, bounds_name, str(bounds_dim))
+
+    @classmethod
+    def concat(
+        cls,
+        indexes: Sequence[IntervalIndex],
+        dim: Hashable,
+        positions: Iterable[Iterable[int]] | None = None,
+    ) -> IntervalIndex:
+        new_index = PandasIndex.concat(
+            [idx._index for idx in indexes], dim, positions=positions
+        )
+
+        if indexes:
+            bounds_name = indexes[0]._bounds_name
+            bounds_dim = indexes[0]._bounds_dim
+            if any(
+                idx._bounds_name != bounds_name or idx._bounds_dim != bounds_dim
+                for idx in indexes
+            ):
+                raise ValueError(
+                    f"Cannot concatenate along dimension {dim!r} indexes with different "
+                    "boundary coordinate or dimension names"
+                )
+        else:
+            bounds_name = new_index.index.name + "_bounds"
+            bounds_dim = "bnd"
+
+        return cls(new_index, bounds_name, bounds_dim)
+
+    @property
+    def _pd_index(self) -> pd.IntervalIndex:
+        # For typing purpose only
+        # TODO: cleaner to make PandasIndex a generic class, i.e., PandasIndex[pd.IntervalIndex]
+        # will be easier once PEP 696 is fully supported (starting from Python 3.13)
+        return cast(pd.IntervalIndex, self._index.index)
+
+    def create_variables(
+        self, variables: Mapping[Any, Variable] | None = None
+    ) -> dict[Any, Variable]:
+        if variables is None:
+            variables = {}
+        empty_var = Variable((), 0)
+        bounds_attrs = variables.get(self._bounds_name, empty_var).attrs
+        mid_attrs = variables.get(self._index.dim, empty_var).attrs
+
+        # TODO: create a PandasIndexingAdapter subclass for the boundary variable
+        # and wrap it here (avoid data copy)
+        bounds_var = Variable(
+            dims=(self._bounds_dim, self._index.dim),
+            data=np.stack([self._pd_index.left, self._pd_index.right], axis=0),
+            attrs=bounds_attrs,
+        )
+        # TODO: use PandasIndexingAdapter directly (avoid data copy)
+        # and/or maybe add an index build option to preserve original labels?
+        # (if those differ from interval midpoints as defined by pd.IntervalIndex)
+        mid_var = Variable(
+            dims=(self._index.dim,),
+            data=self._pd_index.mid,
+            attrs=mid_attrs,
+        )
+
+        return {self._index.dim: mid_var, self._bounds_name: bounds_var}
+
+    def should_add_coord_to_array(
+        self,
+        name: Hashable,
+        var: Variable,
+        dims: set[Hashable],
+    ) -> bool:
+        # add both the mid and boundary coordinates if the index dimension
+        # is present in the array dimensions
+        if self._index.dim in dims:
+            return True
+        else:
+            return False
+
+    def to_pandas_index(self) -> pd.Index:
```

Comment:
This isn't strictly correct if the central value is not the mid-point, shall we error instead?

Reply:
Yes. I simply removed it since by default `Index.to_pandas_index()` raises an error.